compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -152,14 +156,22 @@ trait TypeOps { this: Context => // TODO: Make standalone object.
         defn.ObjectClass :: Nil
     }
 
-    def mergeRefined(tp1: Type, tp2: Type): Type = {
+    def mergeRefinedOrApplied(tp1: Type, tp2: Type): Type = {
       def fail = throw new AssertionError(i"Failure to join alternatives $tp1 and $tp2")
       tp1 match {
         case tp1 @ RefinedType(parent1, name1, rinfo1) =>
           tp2 match {
             case RefinedType(parent2, `name1`, rinfo2) =>
               tp1.derivedRefinedType(
-                mergeRefined(parent1, parent2), name1, rinfo1 | rinfo2)
+                mergeRefinedOrApplied(parent1, parent2), name1, rinfo1 | rinfo2)
+            case _ => fail
+          }
+        case tp1 @ AppliedType(tycon1, args1) =>
```

Comment:
Complication: need to duplicate merge logic between RefinedTypes and AppliedTypes